README.md
# analytics.js boilerplate

Best practices for implementing analytics.js on modern websites.

For an in-depth explanation of all the features used in this boilerplate (as well as how to report on them), see my article:

[**The Google Analytics Setup I Use on Every Site I Build**](https://philipwalton.com/articles/the-google-analytics-setup-i-use-on-every-site-i-build/) &#8594;

## Boilerplate versions

### [`analytics/base.js`](/src/analytics/base.js)

The base boilerplate extends the [default tracking snippet](https://developers.google.com/analytics/devguides/collection/analyticsjs/#alternative_async_tracking_snippet) and includes the following features:

- Tracks uncaught errors.
- Tracks custom user, session, and hit-level dimensions.
- Sends an initial pageview.
- Sends a pageload performance event.

### [`analytics/autotrack.js`](/src/analytics/autotrack.js)

The autotrack boilerplate builds on top the base boilerplate and includes [select autotrack plugins](https://philipwalton.com/articles/the-google-analytics-setup-i-use-on-every-site-i-build/#using-autotrack-plugins)

### [`analytics/multiple-trackers.js`](/src/analytics/multiple-trackers.js)

The multiple-trackers boilerplate builds on the autotrack boilerplate and includes support for using [multiple trackers](https://philipwalton.com/articles/the-google-analytics-setup-i-use-on-every-site-i-build/#testing-your-implementation).

## Running the boilerplate locally

analytics.js boilerplate uses [webpack](https://webpack.js.org/) to compile the source and [webpack-dev-server](https://github.com/webpack/webpack-dev-server) to run it locally.

To install the dependencies and load the boilerplate in a browser, run the following commands:

```sh
npm install
npm start
```

Then visit [localhost:8080](http://localhost:8080) in your browser and open the developer console to see the analytics.js debug output.

### Running different boilerplate versions

The boilerplate [`index.js`](/src/index.js#L7) JavaScript file imports the base boilerplate by default. To run a different version, replace the URL imported via `import('./analytics/base.js')` with the version you want to load.


CHANGELOG.md
# Changelog

### 0.2.1 (2017-04-06)

- Support `error` events with no `.error` property

### 0.2.0 (2017-03-07)

- Update hit time calculations to account for the `queueTime` field (#10)
- Add more descriptive error tracking (#9)
- Update error tracking event category and action (#9)

### 0.1.0 (2017-02-08)

- Initial public release


test/analytics-test.js
import assert from 'assert';
import sinon from 'sinon';
import qs from 'querystring';
import * as analytics from '../src/analytics';


const CLIENT_ID_PATTERN = /^\d+\.\d+$/;


const UUID_PATTERN =
    /^[\da-f]{8}-[\da-f]{4}-4[\da-f]{3}-[8-9a-b][\da-f]{3}-[\da-f]{12}$/;


describe('analytics', () => {
  describe('init', () => {
    beforeEach(() => {
      // Ensure sendBeacon exists so that transport mechanism is always used.
      if (!navigator.sendBeacon) navigator.sendBeacon = () => true;

      sinon.stub(navigator, 'sendBeacon', () => true);
    });

    afterEach(() => {
      ga('remove');
      navigator.sendBeacon.restore();
    });

    it('creates a new tracker with the proper default fields', (done) => {
      analytics.init();

      // Wait for a pageview and a perf event.
      waitForHits(2).then(() => {
        ga((tracker) => {
          assert.strictEqual(tracker.get('transport'), 'beacon');
          done();
        });
      });
    });

    it('reports any errors that occured before tracker creation', () => {
      // Creates two fake error events, one with an error object like would
      // appear in most browsers, and one without, like would appear in IE9.
      window.__e.q = [{error: new Error('foo')}, {}];

      analytics.init();

      // Waits for a pageview, two script errors, and a perf event.
      return waitForHits(4).then(() => {
        // Just get the events.
        const hits = getHits((params) => params.ec == 'Script');

        assert.equal(hits[0].ec, 'Script');
        assert.equal(hits[0].ea, 'uncaught error');
        assert(hits[0].el.length);

        assert.equal(hits[1].ec, 'Script');
        assert.equal(hits[1].ea, 'uncaught error');
        assert(hits[1].el === '(not set)');

        delete window.__e;
      });
    });

    it('sends custom dimensions with every hit', () => {
      analytics.init();

      // Wait for a pageview and a perf event.
      return waitForHits(2).then(() => {
        const hits = getHits();

        assert.strictEqual(hits[0].cd1, '1');
        assert(CLIENT_ID_PATTERN.test(hits[0].cd2));
        assert(UUID_PATTERN.test(hits[0].cd3));
        assert(UUID_PATTERN.test(hits[0].cd4));
        assert(hits[0].cd5 > (new Date - 1000));
        assert(hits[0].cd6 === 'pageview');
        assert(hits[0].cd7 === 'pageload');
        assert.strictEqual(hits[0].cd8, document.visibilityState);

        assert.strictEqual(hits[1].cd1, '1');
        assert(CLIENT_ID_PATTERN.test(hits[1].cd2));
        assert(UUID_PATTERN.test(hits[1].cd3));
        assert(UUID_PATTERN.test(hits[1].cd4));
        assert(hits[1].cd5 > (new Date - 1000));
        assert(hits[1].cd6 == 'event');
        assert(hits[1].cd7 == '(not set)');
        assert.strictEqual(hits[1].cd8, document.visibilityState);

        // Assert window IDs are the same.
        assert.strictEqual(hits[0].cd3, hits[1].cd3);

        // Assert hit IDs are *not* the same.
        assert.notStrictEqual(hits[0].cd4, hits[1].cd4);
      });
    });

    it('sends an initial pageview with the hit source', () => {
      analytics.init();

      // Wait for a pageview and a perf event.
      return waitForHits(2).then(() => {
        const hits = getHits();

        assert(hits[0].t === 'pageview');
        assert(hits[0].cd7 === 'pageload');
      });
    });

    it('sends a performance event with navigation timing data', () => {
      analytics.init();

      // Wait for a pageview and a perf event.
      return waitForHits(2).then(() => {
        const hits = getHits();

        assert(/^\d+$/.test(hits[1].cm1));
        assert(/^\d+$/.test(hits[1].cm2));
        assert(/^\d+$/.test(hits[1].cm3));
        assert(parseInt(hits[1].cm2) >= parseInt(hits[1].cm1));
        assert(parseInt(hits[1].cm3) >= parseInt(hits[1].cm2));


        // Assert custom metrics aren't set on the pageview.
        assert.strictEqual(hits[0].cm1, undefined);
        assert.strictEqual(hits[0].cm2, undefined);
        assert.strictEqual(hits[0].cm3, undefined);
      });
    });

    it('sends the performance hit as a non-interaction event', () => {
      analytics.init();

      // Wait for a pageview and a perf event.
      return waitForHits(2).then(() => {
        const hits = getHits();

        assert.strictEqual(hits[1].t, 'event');
        assert.strictEqual(hits[1].ni, '1');
      });
    });
  });
});


const getHits = (filter) => {
  const hits = navigator.sendBeacon.args.map((args) => qs.parse(args[1]));
  return filter ? hits.filter(filter) : hits;
};


const waitForHits = (count) => {
  const startTime = +new Date;
  return new Promise((resolve, reject) => {
    (function f() {
      if (navigator.sendBeacon.callCount === count) {
        resolve();
      } else if (new Date - startTime > 2000) {
        reject(new Error(`Timed out waiting for ${count} hits ` +
            `(${navigator.sendBeacon.callCount} hits received).`));
      } else {
        setTimeout(() => f(count), 100);
      }
    }());
  });
};


test/analytics/base-test.js
import assert from 'assert';
import sinon from 'sinon';
import qs from 'querystring';
import * as analytics from '../../src/analytics/base';


const CLIENT_ID_PATTERN = /^\d+\.\d+$/;


const UUID_PATTERN =
    /^[\da-f]{8}-[\da-f]{4}-4[\da-f]{3}-[8-9a-b][\da-f]{3}-[\da-f]{12}$/;


describe('analytics/base', () => {
  describe('init', () => {
    beforeEach(() => {
      // Ensure sendBeacon exists so that transport mechanism is always used.
      if (!navigator.sendBeacon) navigator.sendBeacon = () => true;

      sinon.stub(navigator, 'sendBeacon').returns(true);
    });

    afterEach(() => {
      ga('remove');
      navigator.sendBeacon.restore();
    });

    it('creates a new tracker with the proper default fields', (done) => {
      analytics.init();

      // Wait for a pageview and a perf event.
      waitForHits(2).then(() => {
        ga((tracker) => {
          assert.strictEqual(tracker.get('transport'), 'beacon');
          done();
        });
      });
    });

    it('reports any errors that occured before tracker creation', () => {
      // Creates two fake error events, one with an error object like would
      // appear in most browsers, and one without, like would appear in IE9.
      window.__e.q = [
        {error: new TypeError('foo')},
        {error: {name: 'ReferenceError', message: 'zomg!'}},
      ];

      analytics.init();

      // Waits for a pageview, two script errors, and a perf event.
      return waitForHits(4).then(() => {
        // Just get the events.
        const hits = getHits((params) => params.ec == 'Uncaught Error');

        assert.strictEqual(hits[0].ec, 'Uncaught Error');
        assert.strictEqual(hits[0].ea, 'TypeError');
        assert(hits[0].el.length);

        assert.strictEqual(hits[1].ec, 'Uncaught Error');
        assert.strictEqual(hits[1].ea, 'ReferenceError');
        assert.strictEqual(hits[1].el, 'zomg!\n(no stack trace)');

        window.__e.q = null;
      });
    });

    it('sends custom dimensions with every hit', () => {
      analytics.init();

      // Wait for a pageview and a perf event.
      return waitForHits(2).then(() => {
        const hits = getHits();

        assert.strictEqual(hits[0].cd1, '1');
        assert(CLIENT_ID_PATTERN.test(hits[0].cd2));
        assert(UUID_PATTERN.test(hits[0].cd3));
        assert(UUID_PATTERN.test(hits[0].cd4));
        assert(hits[0].cd5 > (new Date - 1000));
        assert.strictEqual(hits[0].cd6, 'pageview');
        assert.strictEqual(hits[0].cd7, 'pageload');
        assert.strictEqual(hits[0].cd8, document.visibilityState);

        assert.strictEqual(hits[1].cd1, '1');
        assert(CLIENT_ID_PATTERN.test(hits[1].cd2));
        assert(UUID_PATTERN.test(hits[1].cd3));
        assert(UUID_PATTERN.test(hits[1].cd4));
        assert(hits[1].cd5 > (new Date - 1000));
        assert.strictEqual(hits[1].cd6, 'event');
        assert.strictEqual(hits[1].cd7, '(not set)');
        assert.strictEqual(hits[1].cd8, document.visibilityState);

        // Assert window IDs are the same.
        assert.strictEqual(hits[0].cd3, hits[1].cd3);

        // Assert hit IDs are *not* the same.
        assert.notStrictEqual(hits[0].cd4, hits[1].cd4);
      });
    });

    it('accounts for the queueTime field in hit time calculations', () => {
      analytics.init();
      ga('send', 'pageview', {queueTime: 60 * 60 * 1000});

      // Wait for a pageview, a perf event, and another pageview
      return waitForHits(3).then(() => {
        const hits = getHits();

        assert(+hits[2].cd5 < +hits[0].cd5);
      });
    });

    it('sends an initial pageview with the hit source', () => {
      analytics.init();

      // Wait for a pageview and a perf event.
      return waitForHits(2).then(() => {
        const hits = getHits();

        assert.strictEqual(hits[0].t, 'pageview');
        assert.strictEqual(hits[0].cd7, 'pageload');
      });
    });

    it('sends a performance event with navigation timing data', () => {
      analytics.init();

      // Wait for a pageview and a perf event.
      return waitForHits(2).then(() => {
        const hits = getHits();

        assert(/^\d+$/.test(hits[1].cm1));
        assert(/^\d+$/.test(hits[1].cm2));
        assert(/^\d+$/.test(hits[1].cm3));
        assert(parseInt(hits[1].cm2) >= parseInt(hits[1].cm1));
        assert(parseInt(hits[1].cm3) >= parseInt(hits[1].cm2));


        // Assert custom metrics aren't set on the pageview.
        assert.strictEqual(hits[0].cm1, undefined);
        assert.strictEqual(hits[0].cm2, undefined);
        assert.strictEqual(hits[0].cm3, undefined);
      });
    });

    it('sends the performance hit as a non-interaction event', () => {
      analytics.init();

      // Wait for a pageview and a perf event.
      return waitForHits(2).then(() => {
        const hits = getHits();

        assert.strictEqual(hits[1].t, 'event');
        assert.strictEqual(hits[1].ec, 'Navigation Timing');
        assert.strictEqual(hits[1].ea, 'track');
        assert.strictEqual(hits[1].el, '(not set)');
        assert.strictEqual(hits[1].ni, '1');
      });
    });
  });
});


const getHits = (filter) => {
  const hits = navigator.sendBeacon.args.map((args) => qs.parse(args[1]));
  return filter ? hits.filter(filter) : hits;
};


const waitForHits = (count) => {
  const startTime = +new Date;
  return new Promise((resolve, reject) => {
    (function f() {
      if (navigator.sendBeacon.callCount === count) {
        resolve();
      } else if (new Date - startTime > 2000) {
        reject(new Error(`Timed out waiting for ${count} hits ` +
            `(${navigator.sendBeacon.callCount} hits received).`));
      } else {
        setTimeout(() => f(count), 100);
      }
    }());
  });
};


test/analytics/index-test.js
import assert from 'assert';
import sinon from 'sinon';
import qs from 'querystring';
import * as analytics from '../../src/analytics';


const CLIENT_ID_PATTERN = /^\d+\.\d+$/;


const UUID_PATTERN =
    /^[\da-f]{8}-[\da-f]{4}-4[\da-f]{3}-[8-9a-b][\da-f]{3}-[\da-f]{12}$/;


describe('analytics', () => {
  describe('init', () => {
    beforeEach(() => {
      // Ensure sendBeacon exists so that transport mechanism is always used.
      if (!navigator.sendBeacon) navigator.sendBeacon = () => true;

      sinon.stub(navigator, 'sendBeacon', () => true);
    });

    afterEach(() => {
      ga('remove');
      navigator.sendBeacon.restore();
    });

    it('creates a new tracker with the proper default fields', (done) => {
      analytics.init();

      // Wait for a pageview and a perf event.
      waitForHits(2).then(() => {
        ga((tracker) => {
          assert.strictEqual(tracker.get('transport'), 'beacon');
          done();
        });
      });
    });

    it('reports any errors that occured before tracker creation', () => {
      // Creates two fake error events, one with an error object like would
      // appear in most browsers, and one without, like would appear in IE9.
      window.__e.q = [{error: new Error('foo')}, {}];

      analytics.init();

      // Waits for a pageview, two script errors, and a perf event.
      return waitForHits(4).then(() => {
        // Just get the events.
        const hits = getHits((params) => params.ec == 'Script');

        assert.strictEqual(hits[0].ec, 'Script');
        assert.strictEqual(hits[0].ea, 'uncaught error');
        assert(hits[0].el.length);

        assert.strictEqual(hits[1].ec, 'Script');
        assert.strictEqual(hits[1].ea, 'uncaught error');
        assert.strictEqual(hits[1].el, '(not set)');

        window.__e.q = null;
      });
    });

    it('sends custom dimensions with every hit', () => {
      analytics.init();

      // Wait for a pageview and a perf event.
      return waitForHits(2).then(() => {
        const hits = getHits();

        assert.strictEqual(hits[0].cd1, '1');
        assert(CLIENT_ID_PATTERN.test(hits[0].cd2));
        assert(UUID_PATTERN.test(hits[0].cd3));
        assert(UUID_PATTERN.test(hits[0].cd4));
        assert(hits[0].cd5 > (new Date - 1000));
        assert.strictEqual(hits[0].cd6, 'pageview');
        assert.strictEqual(hits[0].cd7, 'pageload');
        assert.strictEqual(hits[0].cd8, document.visibilityState);

        assert.strictEqual(hits[1].cd1, '1');
        assert(CLIENT_ID_PATTERN.test(hits[1].cd2));
        assert(UUID_PATTERN.test(hits[1].cd3));
        assert(UUID_PATTERN.test(hits[1].cd4));
        assert(hits[1].cd5 > (new Date - 1000));
        assert.strictEqual(hits[1].cd6, 'event');
        assert.strictEqual(hits[1].cd7, '(not set)');
        assert.strictEqual(hits[1].cd8, document.visibilityState);

        // Assert window IDs are the same.
        assert.strictEqual(hits[0].cd3, hits[1].cd3);

        // Assert hit IDs are *not* the same.
        assert.notStrictEqual(hits[0].cd4, hits[1].cd4);
      });
    });

    it('sends an initial pageview with the hit source', () => {
      analytics.init();

      // Wait for a pageview and a perf event.
      return waitForHits(2).then(() => {
        const hits = getHits();

        assert.strictEqual(hits[0].t, 'pageview');
        assert.strictEqual(hits[0].cd7, 'pageload');
      });
    });

    it('sends a performance event with navigation timing data', () => {
      analytics.init();

      // Wait for a pageview and a perf event.
      return waitForHits(2).then(() => {
        const hits = getHits();

        assert(/^\d+$/.test(hits[1].cm1));
        assert(/^\d+$/.test(hits[1].cm2));
        assert(/^\d+$/.test(hits[1].cm3));
        assert(parseInt(hits[1].cm2) >= parseInt(hits[1].cm1));
        assert(parseInt(hits[1].cm3) >= parseInt(hits[1].cm2));


        // Assert custom metrics aren't set on the pageview.
        assert.strictEqual(hits[0].cm1, undefined);
        assert.strictEqual(hits[0].cm2, undefined);
        assert.strictEqual(hits[0].cm3, undefined);
      });
    });

    it('sends the performance hit as a non-interaction event', () => {
      analytics.init();

      // Wait for a pageview and a perf event.
      return waitForHits(2).then(() => {
        const hits = getHits();

        assert.strictEqual(hits[1].t, 'event');
        assert.strictEqual(hits[1].ni, '1');
      });
    });
  });
});


const getHits = (filter) => {
  const hits = navigator.sendBeacon.args.map((args) => qs.parse(args[1]));
  return filter ? hits.filter(filter) : hits;
};


const waitForHits = (count) => {
  const startTime = +new Date;
  return new Promise((resolve, reject) => {
    (function f() {
      if (navigator.sendBeacon.callCount === count) {
        resolve();
      } else if (new Date - startTime > 2000) {
        reject(new Error(`Timed out waiting for ${count} hits ` +
            `(${navigator.sendBeacon.callCount} hits received).`));
      } else {
        setTimeout(() => f(count), 100);
      }
    }());
  });
};


src/analytics/base.js
/* global ga */


/**
 * The tracking ID for your Google Analytics property.
 * https://support.google.com/analytics/answer/1032385
 */
const TRACKING_ID = 'UA-12345-1';


/**
 * Bump this when making backwards incompatible changes to the tracking
 * implementation. This allows you to create a segment or view filter
 * that isolates only data captured with the most recent tracking changes.
 */
const TRACKING_VERSION = '1';


/**
 * A default value for dimensions so unset values always are reported as
 * something. This is needed since Google Analytics will drop empty dimension
 * values in reports.
 */
const NULL_VALUE = '(not set)';


/**
 * A mapping between custom dimension names and their indexes.
 */
const dimensions = {
  TRACKING_VERSION: 'dimension1',
  CLIENT_ID: 'dimension2',
  WINDOW_ID: 'dimension3',
  HIT_ID: 'dimension4',
  HIT_TIME: 'dimension5',
  HIT_TYPE: 'dimension6',
  HIT_SOURCE: 'dimension7',
  VISIBILITY_STATE: 'dimension8',
};


/**
 * A mapping between custom metric names and their indexes.
 */
const metrics = {
  RESPONSE_END_TIME: 'metric1',
  DOM_LOAD_TIME: 'metric2',
  WINDOW_LOAD_TIME: 'metric3',
};


/**
 * Initializes all the analytics setup. Creates trackers and sets initial
 * values on the trackers.
 */
export const init = () => {
  // Initialize the command queue in case analytics.js hasn't loaded yet.
  window.ga = window.ga || ((...args) => (ga.q = ga.q || []).push(args));

  createTracker();
  trackErrors();
  trackCustomDimensions();
  sendInitialPageview();
  sendNavigationTimingMetrics();
};


/**
 * Tracks a JavaScript error with optional fields object overrides.
 * This function is exported so it can be used in other parts of the codebase.
 * E.g.:
 *
 *    `fetch('/api.json').catch(trackError);`
 *
 * @param {(Error|Object)=} err
 * @param {Object=} fieldsObj
 */
export const trackError = (err = {}, fieldsObj = {}) => {
  ga('send', 'event', Object.assign({
    eventCategory: 'Error',
    eventAction: err.name || '(no error name)',
    eventLabel: `${err.message}\n${err.stack || '(no stack trace)'}`,
    nonInteraction: true,
  }, fieldsObj));
};


/**
 * Creates the trackers and sets the default transport and tracking
 * version fields. In non-production environments it also logs hits.
 */
const createTracker = () => {
  ga('create', TRACKING_ID, 'auto');

  // Ensures all hits are sent via `navigator.sendBeacon()`.
  ga('set', 'transport', 'beacon');
};


/**
 * Tracks any errors that may have occured on the page prior to analytics being
 * initialized, then adds an event handler to track future errors.
 */
const trackErrors = () => {
  // Errors that have occurred prior to this script running are stored on
  // `window.__e.q`, as specified in `index.html`.
  const loadErrorEvents = window.__e && window.__e.q || [];

  const trackErrorEvent = (event) => {
    // Use a different eventCategory for uncaught errors.
    const fieldsObj = {eventCategory: 'Uncaught Error'};

    // Some browsers don't have an error property, so we fake it.
    const err = event.error || {
      message: `${event.message} (${event.lineno}:${event.colno})`,
    };

    trackError(err, fieldsObj);
  };

  // Replay any stored load error events.
  for (let event of loadErrorEvents) {
    trackErrorEvent(event);
  }

  // Add a new listener to track event immediately.
  window.addEventListener('error', trackErrorEvent);
};


/**
 * Sets a default dimension value for all custom dimensions on all trackers.
 */
const trackCustomDimensions = () => {
  // Sets a default dimension value for all custom dimensions to ensure
  // that every dimension in every hit has *some* value. This is necessary
  // because Google Analytics will drop rows with empty dimension values
  // in your reports.
  Object.keys(dimensions).forEach((key) => {
    ga('set', dimensions[key], NULL_VALUE);
  });

  // Adds tracking of dimensions known at page load time.
  ga((tracker) => {
    tracker.set({
      [dimensions.TRACKING_VERSION]: TRACKING_VERSION,
      [dimensions.CLIENT_ID]: tracker.get('clientId'),
      [dimensions.WINDOW_ID]: uuid(),
    });
  });

  // Adds tracking to record each the type, time, uuid, and visibility state
  // of each hit immediately before it's sent.
  ga((tracker) => {
    const originalBuildHitTask = tracker.get('buildHitTask');
    tracker.set('buildHitTask', (model) => {
      const qt = model.get('queueTime') || 0;
      model.set(dimensions.HIT_TIME, String(new Date - qt), true);
      model.set(dimensions.HIT_ID, uuid(), true);
      model.set(dimensions.HIT_TYPE, model.get('hitType'), true);
      model.set(dimensions.VISIBILITY_STATE, document.visibilityState, true);

      originalBuildHitTask(model);
    });
  });
};


/**
 * Sends the initial pageview to Google Analytics.
 */
const sendInitialPageview = () => {
  ga('send', 'pageview', {[dimensions.HIT_SOURCE]: 'pageload'});
};


/**
 * Gets the DOM and window load times and sends them as custom metrics to
 * Google Analytics via an event hit.
 */
const sendNavigationTimingMetrics = () => {
  // Only track performance in supporting browsers.
  if (!(window.performance && window.performance.timing)) return;

  // If the window hasn't loaded, run this function after the `load` event.
  if (document.readyState != 'complete') {
    window.addEventListener('load', sendNavigationTimingMetrics);
    return;
  }

  const nt = performance.timing;
  const navStart = nt.navigationStart;

  const responseEnd = Math.round(nt.responseEnd - navStart);
  const domLoaded = Math.round(nt.domContentLoadedEventStart - navStart);
  const windowLoaded = Math.round(nt.loadEventStart - navStart);

  // In some edge cases browsers return very obviously incorrect NT values,
  // e.g. 0, negative, or future times. This validates values before sending.
  const allValuesAreValid = (...values) => {
    return values.every((value) => value > 0 && value < 6e6);
  };

  if (allValuesAreValid(responseEnd, domLoaded, windowLoaded)) {
    ga('send', 'event', {
      eventCategory: 'Navigation Timing',
      eventAction: 'track',
      eventLabel: NULL_VALUE,
      nonInteraction: true,
      [metrics.RESPONSE_END_TIME]: responseEnd,
      [metrics.DOM_LOAD_TIME]: domLoaded,
      [metrics.WINDOW_LOAD_TIME]: windowLoaded,
    });
  }
};


/**
 * Generates a UUID.
 * https://gist.github.com/jed/982883
 * @param {string|undefined=} a
 * @return {string}
 */
const uuid = function b(a) {
  return a ? (a ^ Math.random() * 16 >> a / 4).toString(16) :
      ([1e7] + -1e3 + -4e3 + -8e3 + -1e11).replace(/[018]/g, b);
};


src/analytics/multiple-trackers.js
// Import the individual autotrack plugins you want to use.
import 'autotrack/lib/plugins/clean-url-tracker';
import 'autotrack/lib/plugins/max-scroll-tracker';
import 'autotrack/lib/plugins/outbound-link-tracker';
import 'autotrack/lib/plugins/page-visibility-tracker';
import 'autotrack/lib/plugins/url-change-tracker';


/* global ga */


/**
 * Bump this when making backwards incompatible changes to the tracking
 * implementation. This allows you to create a segment or view filter
 * that isolates only data captured with the most recent tracking changes.
 */
const TRACKING_VERSION = '1';


/**
 * A global list of tracker object, including the name and tracking ID.
 * https://support.google.com/analytics/answer/1032385
 */
const ALL_TRACKERS = [
  {name: 'prod', trackingId: 'UA-XXXXX-Y'},
  {name: 'test', trackingId: 'UA-XXXXX-Z'},
];


/**
 * Just the trackers with a name matching `prod`. Using an array filter
 * allows you to have more than one prod tracker if needed.
 */
const PROD_TRACKERS = ALL_TRACKERS.filter(({name}) => /prod/.test(name));


/**
 * Just the trackers with a name matching `test`. Using an array filter
 * allows you to have more than one test tracker if needed.
 */
const TEST_TRACKERS = ALL_TRACKERS.filter(({name}) => /test/.test(name));


/**
 * A default value for dimensions so unset values always are reported as
 * something. This is needed since Google Analytics will drop empty dimension
 * values in reports.
 */
const NULL_VALUE = '(not set)';


/**
 * A mapping between custom dimension names and their indexes.
 */
const dimensions = {
  TRACKING_VERSION: 'dimension1',
  CLIENT_ID: 'dimension2',
  WINDOW_ID: 'dimension3',
  HIT_ID: 'dimension4',
  HIT_TIME: 'dimension5',
  HIT_TYPE: 'dimension6',
  HIT_SOURCE: 'dimension7',
  VISIBILITY_STATE: 'dimension8',
  URL_QUERY_PARAMS: 'dimension9',
};


/**
 * A mapping between custom metric names and their indexes.
 */
const metrics = {
  RESPONSE_END_TIME: 'metric1',
  DOM_LOAD_TIME: 'metric2',
  WINDOW_LOAD_TIME: 'metric3',
  PAGE_VISIBLE: 'metric4',
  MAX_SCROLL_PERCENTAGE: 'metric5',
  PAGE_LOADS: 'metric6',
};


/**
 * Creates a ga() proxy function that calls commands on all passed trackers.
 * @param {!Array} trackers an array or objects containing the `name` and
 *     `trackingId` fields.
 * @return {!Function} The proxied ga() function.
 */
const createGaProxy = (trackers) => {
  return (command, ...args) => {
    for (let {name} of trackers) {
      if (typeof command == 'function') {
        ga(() => {
          command(ga.getByName(name));
        });
      } else {
        ga(`${name}.${command}`, ...args);
      }
    }
  };
};


/**
 * Command queue proxies
 * (exported so they can be called by other modules if needed).
 */
export const gaAll = createGaProxy(ALL_TRACKERS);
export const gaProd = createGaProxy(PROD_TRACKERS);
export const gaTest = createGaProxy(TEST_TRACKERS);


/**
 * Initializes all the analytics setup. Creates trackers and sets initial
 * values on the trackers.
 */
export const init = () => {
  // Initialize the command queue in case analytics.js hasn't loaded yet.
  window.ga = window.ga || ((...args) => (ga.q = ga.q || []).push(args));

  createTrackers();
  trackErrors();
  trackCustomDimensions();
  requireAutotrackPlugins();
  sendNavigationTimingMetrics();
};


/**
 * Creates the trackers and sets the default transport and tracking
 * version fields. In non-production environments it also logs hits.
 */
const createTrackers = () => {
  for (let tracker of ALL_TRACKERS) {
    window.ga('create', tracker.trackingId, 'auto', tracker.name);
  }

  // Ensures all hits are sent via `navigator.sendBeacon()`.
  gaAll('set', 'transport', 'beacon');
};


/**
 * Tracks a JavaScript error with optional fields object overrides.
 * This function is exported so it can be used in other parts of the codebase.
 * E.g.:
 *
 *    `fetch('/api.json').catch(trackError);`
 *
 * @param {(Error|Object)=} err
 * @param {Object=} fieldsObj
 */
export const trackError = (err = {}, fieldsObj = {}) => {
  gaAll('send', 'event', Object.assign({
    eventCategory: 'Error',
    eventAction: err.name || '(no error name)',
    eventLabel: `${err.message}\n${err.stack || '(no stack trace)'}`,
    nonInteraction: true,
  }, fieldsObj));
};


/**
 * Tracks any errors that may have occured on the page prior to analytics being
 * initialized, then adds an event handler to track future errors.
 */
const trackErrors = () => {
  // Errors that have occurred prior to this script running are stored on
  // `window.__e.q`, as specified in `index.html`.
  const loadErrorEvents = window.__e && window.__e.q || [];

  const trackErrorEvent = (event) => {
    // Use a different eventCategory for uncaught errors.
    const fieldsObj = {eventCategory: 'Uncaught Error'};

    // Some browsers don't have an error property, so we fake it.
    const err = event.error || {
      message: `${event.message} (${event.lineno}:${event.colno})`,
    };

    trackError(err, fieldsObj);
  };

  // Replay any stored load error events.
  for (let event of loadErrorEvents) {
    trackErrorEvent(event);
  }

  // Add a new listener to track event immediately.
  window.addEventListener('error', trackErrorEvent);
};


/**
 * Sets a default dimension value for all custom dimensions on all trackers.
 */
const trackCustomDimensions = () => {
  // Sets a default dimension value for all custom dimensions to ensure
  // that every dimension in every hit has *some* value. This is necessary
  // because Google Analytics will drop rows with empty dimension values
  // in your reports.
  Object.keys(dimensions).forEach((key) => {
    gaAll('set', dimensions[key], NULL_VALUE);
  });

  // Adds tracking of dimensions known at page load time.
  gaAll((tracker) => {
    tracker.set({
      [dimensions.TRACKING_VERSION]: TRACKING_VERSION,
      [dimensions.CLIENT_ID]: tracker.get('clientId'),
      [dimensions.WINDOW_ID]: uuid(),
    });
  });

  // Adds tracking to record each the type, time, uuid, and visibility state
  // of each hit immediately before it's sent.
  gaAll((tracker) => {
    const originalBuildHitTask = tracker.get('buildHitTask');
    tracker.set('buildHitTask', (model) => {
      const qt = model.get('queueTime') || 0;
      model.set(dimensions.HIT_TIME, String(new Date - qt), true);
      model.set(dimensions.HIT_ID, uuid(), true);
      model.set(dimensions.HIT_TYPE, model.get('hitType'), true);
      model.set(dimensions.VISIBILITY_STATE, document.visibilityState, true);

      originalBuildHitTask(model);
    });
  });
};


/**
 * Requires select autotrack plugins and initializes each one with its
 * respective configuration options. As an example of using multiple
 * trackers, this function only requires the `maxScrollTracker` and
 * `pageVisibilityTracker` plugins on the test trackers, so you can ensure the
 * data collected is relevant prior to sending it to your production property.
 */
const requireAutotrackPlugins = () => {
  gaAll('require', 'cleanUrlTracker', {
    stripQuery: true,
    queryDimensionIndex: getDefinitionIndex(dimensions.URL_QUERY_PARAMS),
    trailingSlash: 'remove',
  });
  gaTest('require', 'maxScrollTracker', {
    sessionTimeout: 30,
    timeZone: 'America/Los_Angeles',
    maxScrollMetricIndex: getDefinitionIndex(metrics.MAX_SCROLL_PERCENTAGE),
  });
  gaAll('require', 'outboundLinkTracker', {
    events: ['click', 'contextmenu'],
  });
  gaAll('require', 'pageVisibilityTracker', {
    sendInitialPageview: true,
    pageLoadsMetricIndex: getDefinitionIndex(metrics.PAGE_LOADS),
    visibleMetricIndex: getDefinitionIndex(metrics.PAGE_VISIBLE),
    sessionTimeout: 30,
    timeZone: 'America/Los_Angeles',
    fieldsObj: {[dimensions.HIT_SOURCE]: 'pageVisibilityTracker'},
  });
  gaAll('require', 'urlChangeTracker', {
    fieldsObj: {[dimensions.HIT_SOURCE]: 'urlChangeTracker'},
  });
};


/**
 * Gets the DOM and window load times and sends them as custom metrics to
 * Google Analytics via an event hit.
 */
const sendNavigationTimingMetrics = () => {
  // Only track performance in supporting browsers.
  if (!(window.performance && window.performance.timing)) return;

  // If the window hasn't loaded, run this function after the `load` event.
  if (document.readyState != 'complete') {
    window.addEventListener('load', sendNavigationTimingMetrics);
    return;
  }

  const nt = performance.timing;
  const navStart = nt.navigationStart;

  const responseEnd = Math.round(nt.responseEnd - navStart);
  const domLoaded = Math.round(nt.domContentLoadedEventStart - navStart);
  const windowLoaded = Math.round(nt.loadEventStart - navStart);

  // In some edge cases browsers return very obviously incorrect NT values,
  // e.g. 0, negative, or future times. This validates values before sending.
  const allValuesAreValid = (...values) => {
    return values.every((value) => value > 0 && value < 6e6);
  };

  if (allValuesAreValid(responseEnd, domLoaded, windowLoaded)) {
    gaProd('send', 'event', {
      eventCategory: 'Navigation Timing',
      eventAction: 'track',
      eventLabel: NULL_VALUE,
      nonInteraction: true,
      [metrics.RESPONSE_END_TIME]: responseEnd,
      [metrics.DOM_LOAD_TIME]: domLoaded,
      [metrics.WINDOW_LOAD_TIME]: windowLoaded,
    });
  }
};


/**
 * Accepts a custom dimension or metric and returns it's numerical index.
 * @param {string} definition The definition string (e.g. 'dimension1').
 * @return {number} The definition index.
 */
const getDefinitionIndex = (definition) => +/\d+$/.exec(definition)[0];


/**
 * Generates a UUID.
 * https://gist.github.com/jed/982883
 * @param {string|undefined=} a
 * @return {string}
 */
const uuid = function b(a) {
  return a ? (a ^ Math.random() * 16 >> a / 4).toString(16) :
      ([1e7] + -1e3 + -4e3 + -8e3 + -1e11).replace(/[018]/g, b);
};
